core/src/main/java/org/apache/iceberg/rest/HttpRESTClient.java

```diff
@@ -0,0 +1,288 @@
+        .withMessage(message)
+        .withType(type)
+        .build();
+  }
+
+  private static ErrorResponse parseOrCreateErrorResponse(
+      CloseableHttpResponse response, Optional<String> responseBody, Method method, Class<?> responseType) {
+    try {
+      return responseBody.map(ErrorResponseParser::fromJson).orElseGet(() -> buildDefaultErrorResponse(response));
+    } catch (Exception e) {
+      // Only log a warning that we couldn't parse an error response if some response body was expected
+      if (responseType != null) {
+        LOG.warn(
+            "Failed to parse standard error response for {} call expecting {} response. Creating one instead",
+            method, responseType.getSimpleName(), e);
+      }
+
+      return buildDefaultErrorResponse(response);
+    }
+  }
+
+  /**
+   * Method to execute an HTTP request and process the corresponding response.
+   *
+   * @param method       - HTTP method, such as GET, POST, HEAD, etc.
+   * @param path         - URL path to send the request to
+   * @param requestBody  - Content to place in the request body
+   * @param responseType - Class of the Response type. Needs to have serializer registered with ObjectMapper
+   * @param errorHandler - Error handler delegated for HTTP responses which handles server error responses
+   * @param <T>          - Class type of the response for deserialization. Must be registered with the ObjectMapper.
+   * @return The response entity, parsed and converted to its type T
+   */
+  @Nullable
+  public <T> T execute(
+      Method method, String path, Object requestBody, Class<T> responseType, Consumer<ErrorResponse> errorHandler) {
+    if (path.startsWith("/")) {
+      throw new RESTException(
+          "Received a malformed path for a REST request: %s. Paths should not start with /", path);
+    }
+
+    // TODO - Add RESTUtil.stripTrailingSlash call on uri once PR #4244 is merged and update associated RESTClient
+    //        tests that have special slash handling when using mock-server.
+    String fullUri = String.format("%s/%s", uri, path);
+    HttpUriRequestBase request = new HttpUriRequestBase(method.name(), URI.create(fullUri));
+    addRequestHeaders(request);
+
+    if (requestBody != null) {
+      try {
+        StringEntity stringEntity = new StringEntity(mapper.writeValueAsString(requestBody));
+        request.setEntity(stringEntity);
+      } catch (JsonProcessingException e) {
+        throw new RESTException(e, "Failed to write request requestBody: %s", requestBody);
+      }
+    }
+
+    requestInterceptor.accept(request);
+
+    try (CloseableHttpResponse response = httpClient.execute(request)) {
+
+      Optional<String> responseBody = extractResponseBodyAsString(response, responseType);
+
+      if (!isSuccessful(response)) {
+        ErrorResponse errorResponse = parseOrCreateErrorResponse(response, responseBody, method, responseType);
+        errorHandler.accept(errorResponse);
```

Comment:
Return here in case the handler doesn't throw? Or should this throw a default `RESTException`?

Reply:
I threw a `RESTException`. We might want to revisit that, but for now I think that's good.